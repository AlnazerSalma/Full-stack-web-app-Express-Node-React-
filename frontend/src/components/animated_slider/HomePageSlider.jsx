import React, { useRef } from "react";
import "./AnimatedSlider.css";
import WorksSliderPanel from "../right_slider_panel/WorksCareerRightSlidePanel";
import useFetch from "../../utils/useFetch";
import useRightSlidePanel from "../../hooks/useRightSlidePanel";
import useSliderAnimation from "../../hooks/useSliderAnimation";
import mediaRenderer from "../../utils/mediaRenderer";

const HomePageSlider = () => {
  const sliderTrackRef = useRef(null);
  const {
    data: works,
    loading,
    error,
  } = useFetch("http://localhost:5000/api/works");
  const slideDuration = 40;
  const { isOpen, selectedItem, openPanel, closePanel } = useRightSlidePanel();

  useSliderAnimation(sliderTrackRef, works, slideDuration);

  if (loading) {
    return <div className="loading-message">{loading}</div>;
  }

  if (error) {
    return <p className="error-message">Failed to fetch slider data.</p>;
  }

  return (
    <div className="slider-wrapper">
      <div className="slider-track" ref={sliderTrackRef}>
        {works.map((work, index) => (
          <div
            className="slider-item"
            key={index}
            onClick={() => openPanel(work)}
          >
            <div className="slider-content">
              <h3>{work.title || work.name}</h3>
              <p>{work.desc}</p>
            </div>
            {mediaRenderer({
              src: work.media?.[0]?.src,
              type: work.media?.[0]?.type,
              alt: work.title || work.name,
              className: "media-class-name",
            })}
          </div>
        ))}
      </div>
      {selectedItem && (
        <WorksSliderPanel
          isOpen={isOpen}
          onClose={closePanel}
          role={selectedItem}
        />
      )}
    </div>
  );
};

export default HomePageSlider;
