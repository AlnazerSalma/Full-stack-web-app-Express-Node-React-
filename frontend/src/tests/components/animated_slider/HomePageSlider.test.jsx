import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import HomePageSlider from "../../../components/animated_slider/HomePageSlider";
import mockMediaRenderer from "../../__test_utils__/mockMediaRenderer";
import { mockUseFetch, mockUsePanel } from "../../__test_utils__/mockHooks";
import mockWork from "../../__mocks_data__/mockWork";

jest.mock("../../../utils/mediaRenderer", () => ({
  __esModule: true,
  default: require("../../__test_utils__/mockMediaRenderer").default,
}));

jest.mock("../../../hooks/useSliderAnimation");

jest.mock(
  "../../../components/right_slider_panel/WorksRightSlidePanel",
  () => ({ isOpen, selectedItem, openPanel, closePanel, ...rest }) => (
    <div data-testid="works-slider-panel" {...rest}>
      Mocked WorksSliderPanel
    </div>
  )
);

describe("HomePageSlider", () => {
  const mockWorks = [mockWork];
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders loading state and matches snapshot", () => {
    mockUseFetch({
      loading: "Loading works...",
    });
    mockUsePanel({});

   const { container } = render(<HomePageSlider />);

    expect(screen.getByText("Loading works...")).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it("renders error message and matches snapshot", () => {
   mockUseFetch({
      error: true,
    });
    mockUsePanel({});

      const { container } = render(<HomePageSlider />);

    expect(screen.getByText("Failed to fetch slider data.")).toBeInTheDocument();

    expect(container).toMatchSnapshot();

  });

  it("renders slider items and mediaRenderer called correctly and matches snapshot", () => {
    mockUseFetch({
      data: mockWorks,
    });

    const openPanelMock = jest.fn();
    mockUsePanel({ openPanel: openPanelMock });

    const { container } = render(<HomePageSlider />);

    mockWorks.forEach(({ title, name, desc }) => {
      const text = title || name;
      expect(screen.getByText(text)).toBeInTheDocument();
      expect(screen.getByText(desc)).toBeInTheDocument();
    });

    expect(mockMediaRenderer).toHaveBeenCalledTimes(mockWorks.length);
    mockWorks.forEach((work, index) => {
      expect(mockMediaRenderer.mock.calls[index][0]).toEqual(
        expect.objectContaining({
          src: work.media?.[0]?.src,
          type: work.media?.[0]?.type,
          alt: work.title || work.name,
          className: "media-class-name",
        })
      );
    });

    expect(container).toMatchSnapshot();
  });


  it("calls openPanel when slider item clicked", () => {
    mockUseFetch({
      data: mockWorks,
    });
    const openPanelMock = jest.fn();
    mockUsePanel({
      openPanel: openPanelMock,
    });

    render(<HomePageSlider />);

    const firstSliderItem = screen
      .getByText(mockWorks[0].title || mockWorks[0].name)
      // eslint-disable-next-line testing-library/no-node-access
      .closest(".slider-item");
    expect(firstSliderItem).toBeInTheDocument();

    fireEvent.click(firstSliderItem);
    expect(openPanelMock).toHaveBeenCalledWith(mockWorks[0]);
  });

 it("renders WorksSliderPanel when selectedItem exists", async () => {
  mockUseFetch({
    data: mockWorks,
  });

  mockUsePanel({
    isOpen: true,
    selectedItem: mockWorks[0],
  });

  const { container } = render(<HomePageSlider />);

  // eslint-disable-next-line testing-library/prefer-presence-queries
  expect(screen.queryByText("Loading details...")).toBeInTheDocument();
  await waitFor(() => {
    expect(screen.getByTestId("works-slider-panel")).toBeInTheDocument();
    // eslint-disable-next-line testing-library/no-wait-for-multiple-assertions
    expect(screen.getByTestId("works-slider-panel")).toHaveAttribute(
      "role",
      "[object Object]"
    );
  });
  expect(container).toMatchSnapshot();
});

});
