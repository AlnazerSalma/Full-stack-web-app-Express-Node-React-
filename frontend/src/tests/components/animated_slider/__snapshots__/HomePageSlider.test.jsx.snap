// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePageSlider renders WorksSliderPanel when selectedItem exists 1`] = `
<div>
  <div
    class="slider-wrapper"
  >
    <div
      class="slider-track"
    >
      <div
        class="slider-item"
      >
        <div
          class="slider-content"
        >
          <h3>
            Frontend Developer
          </h3>
          <p>
            Responsible for building UI components.
          </p>
        </div>
      </div>
    </div>
    <div
      data-testid="works-slider-panel"
      role="[object Object]"
    >
      Mocked WorksSliderPanel
    </div>
  </div>
</div>
`;

exports[`HomePageSlider renders error message and matches snapshot 1`] = `
<div>
  <p
    class="error-message"
  >
    Failed to fetch slider data.
  </p>
</div>
`;

exports[`HomePageSlider renders loading state and matches snapshot 1`] = `
<div>
  <div
    class="loading-message"
  >
    Loading works...
  </div>
</div>
`;

exports[`HomePageSlider renders slider items and mediaRenderer called correctly and matches snapshot 1`] = `
<div>
  <div
    class="slider-wrapper"
  >
    <div
      class="slider-track"
    >
      <div
        class="slider-item"
      >
        <div
          class="slider-content"
        >
          <h3>
            Frontend Developer
          </h3>
          <p>
            Responsible for building UI components.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
